src/collections/map.rs

```diff
@@ -219,6 +273,123 @@ where
+                (false, true) => {
+                    let entry = store_iter
+                        .next()
+                        .transpose()?
+                        .expect("Peek ensures that this in unreachable");
+
+                    let decoded_key: K = Decode::decode(entry.0.as_slice())?;
+                    let decoded_value: <V as State<S>>::Encoding =
+                        Decode::decode(entry.1.as_slice())?;
+
+                    let value_store = parent_store.sub(entry.0.as_slice());
+
+                    Some((
+                        Ref::Owned(decoded_key),
+                        Ref::Owned(V::create(value_store, decoded_value)?),
+                    ))
+                }
+
+                // merge values from both iterators
+                (true, true) => {
+                    let map_key = map_iter.peek().unwrap().0;
+                    let backing_key = match store_iter.peek().unwrap() {
+                        Err(err) => failure::bail!("{}", err),
+                        Ok((ref key, _)) => key,
+                    };
+
+                    let decoded_backing_key: K = Decode::decode(backing_key.as_slice())?;
+                    let key_cmp = map_key.cmp(&decoded_backing_key);
+
+                    // map_key > backing_key, emit the backing entry
+                    // map_key == backing_key, map entry shadows backing entry
+                    if key_cmp == Ordering::Greater {
+                        let entry = store_iter.next().unwrap()?;
+                        let decoded_key: K = Decode::decode(entry.0.as_slice())?;
+                        let decoded_value: <V as State<S>>::Encoding =
+                            Decode::decode(entry.1.as_slice())?;
+
+                        let value_store = parent_store.sub(entry.0.as_slice());
+                        return Ok(Some((
+                            Ref::Owned(decoded_key),
+                            Ref::Owned(V::create(value_store, decoded_value)?),
+                        )));
+                    }
+
+                    if key_cmp == Ordering::Equal {
+                        store_iter.next();
+                    }
+
+                    // map_key < backing_key
+                    match map_iter.next().unwrap() {
+                        (key, Some(value)) => Some((Ref::Borrowed(key), Ref::Borrowed(value))),
+
+                        // map entry deleted in in-memory map, skip
+                        (_, None) => continue,
+                    }
+                }
+            });
+        }
+    }
+}
+
+impl<'a, K, V, S> Iterator for Iter<'a, K, V, S>
+where
+    K: Next<K> + Decode + Encode + Terminated + Hash + Eq + Ord,
+    V: State<S> + Decode,
+    S: Read,
+{
+    type Item = Result<(Ref<'a, K>, Ref<'a, V>)>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        Iter::iter_merge_next(self.parent_store, &mut self.map_iter, &mut self.store_iter)
```

Comment:
I mentioned that `iter_merge_next` could be a method, I guess it could also just all be here inside `next`, not sure if one way is cleaner.

Reply:
Good point. Now that the iter_merge_next takes an Iter, it doesn't make sense to have item_merge_next at all when self can be referenced directly from the next method. In any case, item_merge_next is already a method on Iter, so I think it's a bit cleaner to have next contain all of the next logic rather than calling item_merge_next and having a separate implementation block.

Reply:
Scratch that. iter_merge_next returns Result<Option<(_, _)>> and next wants that transposed. The single transpose on the returned value from iter_merge_next will be a bit cleaner than the refactored alternative in the next logic.